### PR TITLE
DEV-2230 Design and implement OTA service implementation in OCPP server

### PR DESCRIPTION
**Pull request by rhstr**
Tasks skipped on purpose as I cannot imagine a polling scenario here (reports should be triggered manually by calling a service method)

### CODE REVIEW

File: adapter/service/ota/service.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+	"github.com/futurehomeno/fimpgo"
+	"github.com/futurehomeno/fimpgo/fimptype"
+
+	"github.com/futurehomeno/cliffhanger/adapter"
+)
+
+// Controller is an interface representing a device capable of OTA firmware updated.
+// In a polling scenario implementation might require some safeguards against excessive polling.
+type Controller interface {
```
Comment (boguslaw-wojcik) [inline adapter/service/ota/service.go:15]: If we need to we might split this controller in future into two, one for starting, which is optional, and one for reporting which is mandatory. Zigbee for instance does its OTA on its own and does not allow external trigger. But it's just a remark for future.

File: adapter/service/ota/service.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	if err = s.SendMessage(message); err != nil {
+		return fmt.Errorf("failed to send OTA status report: %w", err)
+	}
+
+	return nil
+}
+
+func newProgressReport(data ProgressData) *fimpgo.FimpMessage {
```
Comment (boguslaw-wojcik) [inline adapter/service/ota/service.go:124]: It's not a requirement, just a remark noting a preference. By default I attach as many private methods as possible to the struct of a dedicated service, including all helper methods used by this service. Why? To not pollute a private scope of the package with too many functions. Why? It's easier to work with such packages using IDE, autocompletion and even CoPilot. It does not matter here, just a general remark.

File: adapter/service/ota/service.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	if err = report.validate(); err != nil {
+		return fmt.Errorf("invalid OTA update report: %w", err)
+	}
+
+	var message *fimpgo.FimpMessage
+
+	switch report.Status {
+	case StatusIdle:
+		return nil
+	case StatusInProgress:
```
Comment (boguslaw-wojcik) [inline adapter/service/ota/service.go:111]: Does it mean that the controller is responsible for returning "end report" once and after that only idle?

What about a situation when the OTA update is started by a different process than the FIMP command?

File: adapter/service/ota/service.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	// StartUpdate starts an OTA update process with provided firmware path.
+	StartUpdate(firmwarePath string) error
+
+	// SendStatusReport sends an OTA status report.
+	// If controller reports idle status, no report is sent.
+	SendStatusReport() error
+}
+
+// Config represents a service configuration.
+type Config struct {
```
Comment (boguslaw-wojcik) [inline adapter/service/ota/service.go:36]: I think the service implementation needs to be different. IMHO It needs methods (besides StartUpdate):

SendStartReport
SendProgressReport
SendEndReport

Entire logic of triggering those methods must be on the controller implementation.

There are no getters for these.

**Review by boguslaw-wojcik — CHANGES_REQUESTED** [review]
Logic of the implementation need to shift complete responsibility on the controller and do not keep it in the middle.

File: adapter/service/ota/service.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+// Controller is an interface representing a device capable of OTA firmware updated.
+// In a polling scenario implementation might require some safeguards against excessive polling.
+type Controller interface {
+	// StartOTAUpdate starts an OTA update process with provided firmware path.
+	StartOTAUpdate(firmwarePath string) error
+
+	// OTAUpdateReport returns an OTA update report.
+	OTAUpdateReport() (UpdateReport, error)
```
Comment (boguslaw-wojcik) [inline adapter/service/ota/service.go:20]: OTA Status Report

File: adapter/service/ota/service.go (hunk 2 above)
Comment (rhstr) [reply adapter/service/ota/service.go:124]: Changed.

**Comment by rhstr** [comment]
The feature is adapted according to the guidelines we discussed on yesterday's call.

**Review by boguslaw-wojcik — APPROVED** [review]

**Pull request merged**